src/esp/gfx/PbrImageBasedLighting.h

```diff
@@ -0,0 +1,199 @@
+   * @param[in] equirectangularImageFilename, the name (path included) of the
+   * equirectangular image, that will be converted to a environment cube map
+   */
+  explicit PbrImageBasedLighting(
+      Flags flags,
+      ShaderManager& shaderManager,
+      const std::string& equirectangularImageFilename);
+
+  /**
+   * @brief get the irradiance cube map
+   */
+  CubeMap& getIrradianceMap();
+
+  /**
+   * @brief get the pre-filtered cube map
+   */
+  CubeMap& getPrefilteredMap();
+
+  /**
+   * @brief get the brdf lookup table
+   */
+  Magnum::GL::Texture2D& getBrdfLookupTable();
+
+  /**
+   * @brief get the brdf lookup table
+   * @param[in] equirectangularImageFilename, the name (path included) of the
+   * equirectangular image, that will be converted to a environment cube map
+   */
+  void convertEquirectangularToCubeMap(
+      const std::string& equirectangularImageFilename);
+
+ private:
+  Flags flags_;
+
+  /**
+   * @brief the environment map (e.g., a sky box)
+   */
+  Cr::Containers::Optional<CubeMap> environmentMap_;
+  /**
+   * @brief irradiance cube map (default size of each face: 64 x 64 pixels),
+   * that stores light radiated from the environment. This is for the
+   * indirect diffuse.
+   */
+  Cr::Containers::Optional<CubeMap> irradianceMap_;
+  /**
+   *@brief Pre-filtered environment map that stores specular contribution for
+   * different roughness values and store the results in the mip-map levels of a
+   * cubemap (namely, a mip chain from roughness = 0.0 to roughness = 1.0.)
+   * This is sampled for the indirect specular.
+   */
+  Cr::Containers::Optional<CubeMap> prefilteredMap_;
+  /**
+   * @brief 2D BRDF lookup table, an HDR image (16-bits per channel) that
+   * contains BRDF values for roughness and view angle. This is for the indirect
+   * specular.
+   * See: Karis, Brian. “Real Shading in Unreal Engine 4” (2013).
+   */
+  Cr::Containers::Optional<Magnum::GL::Texture2D> brdfLUT_;
+
+  ShaderManager& shaderManager_;
+  /**
+   * @brief recreate brdf lookup table, irradiance map, prefiltered environment
+   * map
+   */
+  void recreateTextures();
+
+  /** @brief load the brdf LUT from the disk */
+  // TODO: HDR
+  void loadBrdfLookUpTable();
+
+  enum class PrecomputedMapType : uint8_t {
+    IrradianceMap = 0,
+    PrefilteredMap = 1,
+  };
+
+  /**
+   * @brief precompute the irradiance map, prefiltered environment map
+   * @param[in] type, see @ref PrecomputedMapType
+   */
+  void computePrecomputedMap(PrecomputedMapType type);
+
+  enum class PbrIblShaderType : uint8_t {
+    IrradianceMap = 0,
+    PrefilteredMap = 1,
+    // BrdfLookupTable = 2, // TODO
+    EquirectangularToCubeMap = 3,
+  };
+  /**
+   * @brief get the shader based on the type
+   * @param[in] type, see @ref PbrIblShaderType
+   */
+  template <typename T>
+  Mn::Resource<Mn::GL::AbstractShaderProgram, T> getShader(
+      PbrIblShaderType type) {
```

Comment:
This template function is private and only ever used from the cpp file, so its definition could be in the cpp -- no need to have it in the header.